geps/gep-2162.md

```diff
@@ -0,0 +1,114 @@
+# GEP-2162: Supported features in GatewayClass Status
+
+* Issue: [#2162](https://github.com/kubernetes-sigs/gateway-api/issues/2162)
+* Status: Provisional
+
+## TLDR
+
+This GEP proposes to enhance the [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185) to include a list of gw-api features supported by the installed Gateway Class. 
+
+## Goals
+
+* Enable users easily see what features the implementation(gateway class) support
+* Improving UX and potentially provide clear errors when a user is trying to use a feature the installed GWC doesn't support
+
+## Non-Goals
+
+* Validate correctness of supported features published by the implementation.
+
+## Introduction
+
+The current  [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185 is only used to store conditions the controller publishes.
+
+With the evolution of the Conformance tests suite and features, we want to;
+
+1. Enable users easily see what features the implementation support
+1. Improving UX and potentially provide clear errors when a user is trying to use a feature the installed GWC doesn't support
+
+This doc proposes to enhance the GatewayClassStatus API so implementations could publish a list of features they support/don't support.
+
+Implementations **should** publish the supported/unsupported features before Accepting the GatewayClass.
+
+They can either maintain static lists of supported/unsupported features OR generate if from their conformance tests results/reports as part of their CI.
+
+## API
+
+This GEP proposes API changes describes as follow:
+1. Adding `FeaturesSupport` and  `GatewayClassFeatures` structs. 
+2. Update the `GatewayClassStatus` struct to include `GatewayClassFeatures`
+
+```go
+// FeaturesSupport contains the supported features and unsupported features by the installed GatewayClass.
+type FeaturesSupport struct {
+	// SupportedFeatures indicates which extended features were flagged as
+	// supported by the implementation and tests will be attempted for.
+	SupportedFeatures []string `json:"supportedFeatures,omitempty"`
+
+	// UnsupportedFeatures indicates which extended features the implementation
+	// does not have support for and therefore will not attempt to test.
+	UnsupportedFeatures []string `json:"unsupportedFeatures,omitempty"`
```

Comment:
I chose to add both `SupportedFeatures` and `UnsupportedFeatures`, I think it is a better UX if the user also see what features are not supported in his GWC instead of comparing the supported ones with the gw-api documentation to see what's missing.  
Open to debate if anyone has a different opinion

Reply:
I think it's going to be impossible for implementations to keep something like `UnsupportedFeatures` up-to-date. As we get more conformance profiles, I think the list of `UnsupportedFeatures` may become larger than the list of `SupportedFeatures` for some implementations - for example implementations that only support L4 or mesh would have a really long list here. Maybe more problematic, this is something where it's impossible to predict the future, so unless the latest version of an implementation is always deployed, and that implementation always keeps track of the full set of Gateway API features, there are going to be omissions here.

Ideally tooling such as gwctl, webhook, UX, etc should be able to calculate the list of UnsupportedFeatures based on the list of `SupportedFeatures` to help make this UX better, but I don't think we can expect individual implementations to be fully aware of every possible feature that could exist.

Reply:
> I think it's going to be impossible for implementations to keep something like UnsupportedFeatures up-to-date. As we get more conformance profiles, I think the list of UnsupportedFeatures may become larger than the list of SupportedFeatures for some implementations - for example implementations that only support L4 or mesh would have a really long list here. Maybe more problematic, this is something where it's impossible to predict the future, so unless the latest version of an implementation is always deployed, and that implementation always keeps track of the full set of Gateway API features, there are going to be omissions here.

+1

Reply:
done